### PR TITLE
[Inference API] Add Alibaba Cloud AI Search Model support to Inference API

### PR DESCRIPTION
**Pull request by weizijun**
Add Alibaba Cloud AI Search Model support to Inference API.
### Prerequisites to Model Creation
An Alibaba Cloud Account with Alibaba Cloud Opensearch access
An api key used to access Alibaba Cloud AI Search Model
### Inference Model Creation:
```
PUT _inference/{task_type}/{inference_model_id}
{
  "service": "alibabacloud-ai-search",
  "service_settings": {
    "api_key": "{{api_key}}",
    "service_id": "<<service_id>>",
    "host": "<<host>",
    "workspace": "<<workspace_name>>",
    <<ADDITIONAL SERVICE SETTINGS (see below)>>,
  }
  "task_settings": {
    <<TASK SETTINGS (see below)>>
  }
}
```
### Support task_type
We have now added support for these task_types:
- text_embedding
- sparse_embedding
- rerank

We also support other task_types, such as completion, document analyze, document split, image analyze, query analyze.
Here is the document: https://help.aliyun.com/zh/open-search/search-platform/product-overview/introduction-to-search-platform

### Testing
Creating the inference endpoint for Alibaba Cloud AI Search
```
PUT _inference/text_embedding/os-embeddings
{
  "service": "alibabacloud-ai-search",
  "service_settings": {
    "api_key": "{{API_KEY}}",
    "service_id": "ops-text-embedding-001",
    "host" : "default-j01.platform-cn-shanghai.opensearch.aliyuncs.com",
    "workspace" : "default"
  }
}
```
Performing text_embedding inference:
```
POST _inference/text_embedding/os-embeddings
{
    "input": "What is Elastic?"
}
```

### CODE REVIEW

**Comment by saikatsarkar056** [comment]
@davidkyle, @weizijun 

After this backend PR has been merged. We need to add alibaba to inference endpoints view with proper icon and display fields in the front-end. This is an example [PR](https://github.com/elastic/kibana/pull/188670) for  Amazon Bedrock. Thank you.

**Comment by weizijun** [comment]
> After this backend PR has been merged, we need to add alibaba to inference endpoints view with proper icon and display fields in the front-end. This is an example [PR](https://github.com/elastic/kibana/pull/188670) for Amazon Bedrock. Thank you.

Ok, I will try to  add alibaba to inference endpoints view PR. Thank you!
And about this PR, Is there anything else to add to this PR?

**Review by davidkyle — COMMENTED** [review]
This looks really good thank you!

I don't see any problems but this is a large PR so I'd like to review the changes 1 more time and test the inference against an Alibaba AI model myself.

**Comment by davidkyle** [comment]
@weizijun I resolved the conflict in `TransportVersions.java` the CI build should run now

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/alibabacloudsearch/AlibabaCloudSearchSparseRequestEntity.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        if (taskSettings.isReturnToken() != null) {
+            builder.field(RETURN_TOKEN_FIELD, taskSettings.isReturnToken());
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    // default for testing
+    static String covertToString(InputType inputType) {
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/alibabacloudsearch/AlibabaCloudSearchSparseRequestEntity.java:57]: done

**Review by davidkyle — COMMENTED** [review]
Fantastic I was able to create text embeddings, sparse embeddings and rerank documents in Alibaba with this change. 

The comments I've left are all related to the configuration and are suggestions to improve the validation. Otherwise the PR is good Thank You ❤️

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        String url = extractOptionalString(map, URL, ModelConfigurations.SERVICE_SETTINGS, validationException);
+
+        SimilarityMeasure similarity = extractSimilarity(map, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        Integer dims = removeAsType(map, DIMENSIONS, Integer.class);
+        Integer maxInputTokens = removeAsType(map, MAX_INPUT_TOKENS, Integer.class);
+        URI uri = convertToUri(url, URL, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        String modelId = extractRequiredString(map, MODEL_ID, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        String host = extractRequiredString(map, HOST, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        var workspaceName = extractRequiredString(map, WORKSPACE_NAME, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        var httpSchema = extractOptionalString(map, HTTP_SCHEMA_NAME, ModelConfigurations.SERVICE_SETTINGS, validationException);
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java:67]: done

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/rerank/AlibabaCloudSearchRerankTaskSettings.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+ * Defines the task settings for the AlibabaCloudSearch rerank service.
+ *
+ * <p>
+ * <a href="https://help.aliyun.com/zh/open-search/search-platform/developer-reference/ranker-api-details">See api docs for details.</a>
+ * </p>
+ */
+public class AlibabaCloudSearchRerankTaskSettings implements TaskSettings {
+    public static final String NAME = "alibabacloud_search_rerank_task_settings";
+    public static final String RETURN_DOCUMENTS = "return_documents";
+    public static final String TOP_N_DOCS_ONLY = "top_n";
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/rerank/AlibabaCloudSearchRerankTaskSettings.java]: Ok, I remove these fields.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/sparse/AlibabaCloudSearchSparseTaskSettings.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+    public static final String NAME = "alibabacloud_search_sparse_embeddings_task_settings";
+    public static final AlibabaCloudSearchSparseTaskSettings EMPTY_SETTINGS = new AlibabaCloudSearchSparseTaskSettings(null, null);
+    static final String INPUT_TYPE = "input_type";
+    static final String RETURN_TOKEN = "return_token";
+    static final EnumSet<InputType> VALID_REQUEST_VALUES = EnumSet.of(
+        InputType.INGEST,
+        InputType.SEARCH,
+        InputType.CLASSIFICATION,
+        InputType.CLUSTERING
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/sparse/AlibabaCloudSearchSparseTaskSettings.java]: done

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    // todo: 确认开放搜索平台的ratelimt
+    private static final RateLimitSettings DEFAULT_RATE_LIMIT_SETTINGS = new RateLimitSettings(10_000);
+
+    public static AlibabaCloudSearchServiceSettings fromMap(Map<String, Object> map, ConfigurationParseContext context) {
+        ValidationException validationException = new ValidationException();
+
+        String url = extractOptionalString(map, URL, ModelConfigurations.SERVICE_SETTINGS, validationException);
+
+        SimilarityMeasure similarity = extractSimilarity(map, ModelConfigurations.SERVICE_SETTINGS, validationException);
+        Integer dims = removeAsType(map, DIMENSIONS, Integer.class);
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java:61]: yeah, I move these params to AlibabaCloudSearchEmbeddingsServiceSettings

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    public static final String WORKSPACE_NAME = "workspace";
+    public static final String HTTP_SCHEMA_NAME = "http_schema";
+
+    // todo: 确认开放搜索平台的ratelimt
+    private static final RateLimitSettings DEFAULT_RATE_LIMIT_SETTINGS = new RateLimitSettings(10_000);
+
+    public static AlibabaCloudSearchServiceSettings fromMap(Map<String, Object> map, ConfigurationParseContext context) {
+        ValidationException validationException = new ValidationException();
+
+        String url = extractOptionalString(map, URL, ModelConfigurations.SERVICE_SETTINGS, validationException);
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java:52]: I rename this mode_id to service_id, but `modelId()` is a interface method, it return the `service_id`

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+import static org.elasticsearch.xpack.inference.services.ServiceUtils.extractSimilarity;
+import static org.elasticsearch.xpack.inference.services.ServiceUtils.removeAsType;
+
+public class AlibabaCloudSearchServiceSettings extends FilteredXContentObject
+    implements
+        ServiceSettings,
+        AlibabaCloudSearchRateLimitServiceSettings {
+
+    public static final String NAME = "alibabacloud_search_service_settings";
+    public static final String MODEL_ID = "service_id";
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java]: I rename this mode_id to service_id, but `modelId()` is a interface method, it return the `service_id`

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java:52]: I test this case:

```
PUT _inference/text_embedding/osuri
{
  "service": "alibabacloud-ai-search",
  "service_settings": {
    "url" : "https://default-j01.platform-cn-shanghai.opensearch.aliyuncs.com/v3/openapi/workspaces/default/text-embedding/ops-text-embedding-zh-001",
    "api_key": "xxx",
    "service_id": "ops-text-embedding-001",
    "host" : "default-j01.platform-cn-shanghai.opensearch.aliyuncs.com",
    "workspace" : "default"
  }
}
```

It works, and it use the url to replace host\ service_id\ workspace.
The usage of  the `url` parameter is the same as other inference service, so I added it. But maybe we don’t need it.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/alibabacloudsearch/AlibabaCloudSearchEmbeddingsRequestEntity.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        Objects.requireNonNull(input);
+        Objects.requireNonNull(taskSettings);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(TEXTS_FIELD, input);
+        if (taskSettings.getInputType() != null) {
+            builder.field(INPUT_TYPE_FIELD, covertToString(taskSettings.getInputType()));
```
[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/alibabacloudsearch/AlibabaCloudSearchEmbeddingsRequestEntity.java]: ok, I do some refactor.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java:52]: In other inference services the `url` is configurable to support models [deployed with LocalAI](https://www.elastic.co/search-labs/blog/localai-for-text-embeddings). I don't think we need it in this case and it is confusing to have the `url` parameter plus the `host` and `workspace` fields. Please remove the `url` option

[the remark this answers is not in the document]
Comment (weizijun) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/AlibabaCloudSearchServiceSettings.java]: done, I have remove the `url` parameter.

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by davidkyle** [comment]
Thanks @weizijun this is great to have another reranker, sparse embeddings and a text embedding service added. I've asked one of the Elasticsearch writers to create the documentation for this change.

**Comment by weizijun** [comment]
> Thanks @weizijun this is great to have another reranker, sparse embeddings and a text embedding service added. I've asked one of the Elasticsearch writers to create the documentation for this change.

Thanks @davidkyle , What can I do to help with document writing?

**Comment by serenachou** [comment]
@weizijun is there any chance you'd want to add a notebook into https://github.com/elastic/elasticsearch-labs/tree/main/notebooks/integrations to show a few examples? We can use that for a blog post that we can write together too.  there's a few tutorials we can also add Alibaba code examples to : https://www.elastic.co/guide/en/elasticsearch/reference/master/semantic-search-semantic-text.html or https://www.elastic.co/guide/en/elasticsearch/reference/master/semantic-search-inference.html depending on the use case. cc @szabosteve

**Comment by davidkyle** [comment]
@elasticmachine test this please

**Comment by weizijun** [comment]
> @weizijun is there any chance you'd want to add a notebook into https://github.com/elastic/elasticsearch-labs/tree/main/notebooks/integrations to show a few examples?

OK, Let me see, and I will add some examples.

**Comment by davidkyle** [comment]
@elasticmachine test this please

**Comment by davidkyle** [comment]
@elasticmachine test this please

**Comment by weizijun** [comment]
Thanks to @davidkyle for merging this PR. I will write a PR document for our inference service.

**Pull request merged**